keps/sig-node/3857-rro-hostpath/README.md

```diff
@@ -0,0 +1,967 @@
+to make read-only bind mounts literally read-only.
+
+The "rro" bind mount options is implemented by calling [`mount_setattr(2)`](https://man7.org/linux/man-pages/man2/mount_setattr.2.html)
+with `MOUNT_ATTR_RDONLY` and `AT_RECURSIVE`.
+
+Requires runc >= 1.1 && kernel >= 5.12.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The current `readOnly` hostPath volumes are not recursively read-only;
+e.g., even if `/mnt` is mounted as read-only, its submounts such as `/mnt/usbstorage` are not read-only.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Support recursively read-only mounts for runc >= 1.1 && kernel >= 5.12.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursively read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursivelyReadOnly: Required` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursivelyReadOnly: IfPossible`.
```

Comment:
Is `IfPossible` needed for third party tooling that wants universally working solution and requires the fallback option? Typical end user pods will know if the option is working or not.

Or there is some other long term possibility when this setting may or may not work depending on some other factors?

If it is only for thrid party tooling, it can be enforced via some webhook or by providing two Pod definitions and some labels discovery thing like NFD. This way you either enforce it or not enforce and it is clear from the Pod definition which one it is.

Reply:
> Or there is some other long term possibility when this setting may or may not work depending on some other factors?

Doesn't work on old kernel

Reply:
this is my question. If we expect that in a few years there will be no old kernel users any longer, do we need to keep the `IfPossible` option?

Reply:
Not a few years. People will continue to use kernel 4.18 (EL8) until 2029.